main/src/ca/uwaterloo/flix/language/phase/constraintgeneration/TypingContext.scala

```diff
@@ -0,0 +1,209 @@
+    tpes match {
+      case tpe1 :: rest =>
+        rest.foreach(unifyTypeM(tpe1, _, loc))
+        tpe1
+      case Nil => Type.freshVar(kind, loc.asSynthetic)
+    }
+  }
+
+  /**
+    * Generates constraints expecting the given type arguments to unify.
+    */
+  def expectTypeArguments(sym: Symbol, expectedTypes: List[Type], actualTypes: List[Type], actualLocs: List[SourceLocation], loc: SourceLocation)(implicit flix: Flix): Unit = {
+    expectedTypes.zip(actualTypes).zip(actualLocs).zipWithIndex.foreach {
+      case (((expectedType, actualType), loc), index) =>
+        val argNum = index + 1
+        val prov = Provenance.ExpectArgument(expectedType, actualType, sym, argNum, loc)
+        currentScopeInfo.constrs.addOne(TypingConstraint.Equality(expectedType, actualType, prov))
+    }
+  }
+
+  /**
+    * Generates constraints unifying the given types.
+    */
+  def unifyType3M(tpe1: Type, tpe2: Type, tpe3: Type, loc: SourceLocation): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+    unifyTypeM(tpe1, tpe3, loc)
+  }
+
+  /**
+    * Generates constraints unifying the given effects.
+    */
+  // TODO ASSOC-TYPES this should actually do something
+  def unifyEffM(tpe1: Type, tpe2: Type, loc: SourceLocation)(implicit flix: Flix): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+  }
+
+  /**
+    * Generates constraints expecting the given types to unify.
+    */
+  def expectTypeM(expected: Type, actual: Type, loc: SourceLocation): Unit = {
+    currentScopeInfo.constrs.append(TypingConstraint.Equality(expected, actual, Provenance.ExpectType(expected, actual, loc)))
+  }
+
+  /**
+    * Generates constraints expecting the given types to unify, binding them to the bound type.
+    */
+  // TODO ASSOC-TYPES what does this do?
+  def expectTypeBindM(expected: Type, actual: Type, bind: Type, loc: SourceLocation): Unit = {
+    expectTypeM(expected, actual, loc)
+    unifyTypeM(expected, bind, loc)
+  }
+
+  /**
+    * Generates constraints unifying the given Booleans.
+    */
+  // TODO ASSOC-TYPES this should actually do something
+  def unifyBoolM(tpe1: Type, tpe2: Type, loc: SourceLocation): Unit = {
+    unifyTypeM(tpe1, tpe2, loc)
+  }
+
+  /**
+    * Adds the given class constraints to the context.
+    */
+  def addTypeConstraintsM(tconstrs0: List[Ast.TypeConstraint], loc: SourceLocation): Unit = {
+    val tconstrs = tconstrs0.map {
+      case Ast.TypeConstraint(head, arg, _) => TypingConstraint.Class(head.sym, arg, loc)
+    }
+    currentScopeInfo.constrs.addAll(tconstrs)
+  }
+
+  /**
+    * Marks the given type variable as rigid in the context.
+    */
+  def rigidifyM(sym: Symbol.KindedTypeVarSym): Unit = {
+    renv = renv.markRigid(sym)
+  }
+
+  /**
+    * Replaces every occurrence of the effect symbol `sym` with pure in `eff`.
+    *
+    * Note: Does not work for polymorphic effects.
+    */
+  def purifyEff(sym: Symbol.EffectSym, eff: Type): Type = {
```

Comment:
How does this work in the new system? Should it be its own constraint?

Reply:
It doesn't really work. I tried making a dedicated constraint, but it can't work the same way as the purification constraint because the type rule really calls for subtraction, not purification.

This function is an approximation of what's on master, but as you'll see in the constraint generation, we don't follow the role exactly there.